src/node/share_manager.h

```diff
@@ -266,61 +280,72 @@ namespace ccf
+      // wrapped new ledger secret and encrypted previous ledger secret in the
+      // store.
       auto [latest, penultimate] =
         network.ledger_secrets->get_latest_and_penultimate(tx);
 
-      set_recovery_shares_info(tx, latest, penultimate, latest_ls_version);
+      set_recovery_shares_info(tx, latest.second, penultimate, latest.first);
     }
 
-    void issue_shares_on_rekey(
+    void issue_recovery_shares(
       kv::Tx& tx, const LedgerSecret& new_ledger_secret)
     {
+      // Issue new recovery shares of the new ledger secret, recording the
+      // wrapped new ledger secret and encrypted previous ledger secret in the
+      // store.
+      // Note: The version at which the new ledger secret is applicable from is
+      // derived from the hook at which the ledger secret is applied to the
+      // store.
       set_recovery_shares_info(
-        tx, new_ledger_secret, network.ledger_secrets->get_latest(tx).second);
+        tx, new_ledger_secret, network.ledger_secrets->get_latest(tx));
+    }
+
+    void shuffle_recovery_shares(kv::Tx& tx)
+    {
+      // Issue new recovery shares of the _same_ current ledger secret to all
+      // active recovery members. The encrypted ledger secrets recorded in the
+      // store are _not_ updated.
+      _shuffle_recovery_shares(
+        tx, network.ledger_secrets->get_latest(tx).second);
     }
 
     std::optional<EncryptedShare> get_encrypted_share(
       kv::Tx& tx, MemberId member_id)
     {
-      std::optional<EncryptedShare> encrypted_share = std::nullopt;
       auto recovery_shares_info = tx.rw(network.shares)->get(0);
       if (!recovery_shares_info.has_value())
       {
         throw std::logic_error(
           "Failed to retrieve current recovery shares info");
       }
 
-      for (auto const& s : recovery_shares_info->encrypted_shares)
+      auto search = recovery_shares_info->encrypted_shares.find(member_id);
+      if (search == recovery_shares_info->encrypted_shares.end())
       {
-        if (s.first == member_id)
-        {
-          encrypted_share = s.second;
-        }
+        return std::nullopt;
       }
-      return encrypted_share;
+
+      return search->second;
     }
 
     LedgerSecretsMap restore_recovery_shares_info(
-      kv::Tx& tx,
-      const std::list<RecoveredLedgerSecret>& encrypted_recovery_secrets)
+      kv::Tx& tx, RecoveredEncryptedLedgerSecrets&& recovery_ledger_secrets)
     {
       // First, re-assemble the ledger secret wrapping key from the submitted
       // encrypted shares. Then, unwrap the latest ledger secret and use it to
-      // decrypt the previous ledger secret and so on.
+      // decrypt the sequence of recovered ledger secrets, from the last one.
 
-      auto ls_wrapping_key = combine_from_submitted_shares(tx);
+      if (recovery_ledger_secrets.empty())
+      {
+        throw std::logic_error(
+          "There should be at least one encrypted ledger secret to recover");
```

Comment:
"Could not find  any ledger secrets in the ledger"? It seems odd to have that check here, it feel like this is a transform over a sequence and it's more appropriate for the caller to have an opinion about whether empty is a problem.

Reply:
Good point. I've updated the error message.

An empty `recovery_ledger_secrets` list here is an issue as the version of the current ledger secret (in the `shares` table) is indicated by the last entry in `recovery_ledger_secrets`. If no ledger secret were recovered from the ledger, the recovery of the ledger secrets cannot be done. I believe it is the responsibility of this class, rather than the caller.